# INSTRUÇÃO PRÁTICA P001

### Exercício 2: Manipulação de variáveis de tipo inteiro, explorando as características e os limites.


- Operadores aritméticos

In [1]:
# Valores para os cálculos
a = 10
b = 3

# Adição
soma = a + b
print(f"Soma de {a} e {b} é igual a {soma}")

# Subtração
subtracao = a - b
print(f"Subtração de {a} por {b} é igual a {subtracao}")

# Multiplicação
multiplicacao = a * b
print(f"Multiplicação de {a} por {b} é igual a {multiplicacao}")

# Divisão
divisao = a / b
print(f"Divisão de {a} por {b} é igual a {divisao}")

# Módulo
modulo = a % b
print(f"O resto da divisão de {a} por {b} é {modulo}")

# Exponenciação
exponenciacao = a ** b
print(f"{a} elevado a {b} é igual a {exponenciacao}")


Soma de 10 e 3 é igual a 13
Subtração de 10 por 3 é igual a 7
Multiplicação de 10 por 3 é igual a 30
Divisão de 10 por 3 é igual a 3.3333333333333335
O resto da divisão de 10 por 3 é 1
10 elevado a 3 é igual a 1000


- Operadores aritméticos compostos

> Os operadores aritméticos compostos são uma forma simplificada de realizar operações aritméticas e, ao mesmo tempo, atualizar o valor de uma variável. Eles combinam um operador aritmético com uma operação de atribuição.

In [2]:
# Valores iniciais
x = 10
y = 3

# Adição com atribuição
x += y
print(f"Adição com atribuição: x += y resulta em x igual a {x}")

# Subtração com atribuição
x -= y
print(f"Subtração com atribuição: x -= y resulta em x igual a {x}")

# Multiplicação com atribuição
x *= y
print(f"Multiplicação com atribuição: x *= y resulta em x igual a {x}")

# Divisão com atribuição
x /= y
print(f"Divisão com atribuição: x /= y resulta em x igual a {x}")

# Módulo com atribuição
x %= y
print(f"Módulo com atribuição: x %= y resulta em x igual a {x}")

# Exponenciação com atribuição
x **= y
print(f"Exponenciação com atribuição: x **= y resulta em x igual a {x}")


Adição com atribuição: x += y resulta em x igual a 13
Subtração com atribuição: x -= y resulta em x igual a 10
Multiplicação com atribuição: x *= y resulta em x igual a 30
Divisão com atribuição: x /= y resulta em x igual a 10.0
Módulo com atribuição: x %= y resulta em x igual a 1.0
Exponenciação com atribuição: x **= y resulta em x igual a 1.0


- Diferença entre C/C++ : Divisão

> O operador de divisão (/) entre dois inteiros sempre resulta em um número de ponto flutuante em python, mesmo se a divisão for exata. Vimos isso no exemplo acima , quando dividimos 10/3 que deu 3.3333333333333335,enquanto em C/C++, usando inteiros, resultaria em 3. 
>Para obter o mesmo resultado em python é preciso truncar a parte decimal, usando '//'

In [3]:
a = 10
b = 3

# Divisão inteira (comportamento similar ao de C++)
resultado = a // b
print(f"Divisão inteira de {a} por {b} é igual a {resultado}")


Divisão inteira de 10 por 3 é igual a 3


- Diferença entre C/C++ : Exponenciação

> operador de exponenciação é representado por **, enquanto em C/C++ é necessário o uso de funções ou bibliotecas específicas para calcular potências.

In [4]:
# Exponenciação
exponenciacao = a ** b
print(f"{a} elevado a {b} é igual a {exponenciacao}")

10 elevado a 3 é igual a 1000


- Comparação entre limites de representações

> Python permite representar números inteiros significativamente grandes, como fatorial de 30, porém em C/C++ O maior valor positivo que pode ser representado com um int de 32 bits é apenas 2^31 - 1.

In [5]:
import math

# Calculando o fatorial de 30 em Python
fatorial_30 = math.factorial(30)
print(f"O fatorial de 30 em Python é: {fatorial_30}")


O fatorial de 30 em Python é: 265252859812191058636308480000000


In [6]:
import sys

# Maior número inteiro representável em C/C++
maior_int_cpp = 2 ** 31 - 1

print(f"O maior número inteiro representado em C/C++ é: {maior_int_cpp}")


O maior número inteiro representado em C/C++ é: 2147483647


- As variveis são imutaveis, iremos provar atraves do exemplo abaixo

In [7]:
x=64
y=x
x =x + 1 

> Como sabemos, em python tudo é ponteiro, pelo exemplo temos que x e y apontam para o mesmo lugar, intuitivamente podemos achar que ao mudar x também mudaremos Y, mas não é isso que acontece.

In [8]:
print(x)
print(y)

65
64


> x passa a ser 65 mas y continua como 64, pois em python ao atribuir um novo valor a uma variável, ele cria uma nova referência para esse valor (x aponta para um novo endereço), mantendo a referência original inalterada (o endereeço ao qual y aponta não é alterado) e apontando para o mesmo valor anterior. Em linguagens como C, esse novo valor seria armazenado no mesmo endereço de memória onde a variável está alocada.  

- Métodos que estão disponíveis para as variáveis inteiras

> para verificar todos os métodos disponíveis para um tipo de objeto em Python, você pode usar a função dir().

In [9]:
print(dir(int))

['__abs__', '__add__', '__and__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getnewargs__', '__getstate__', '__gt__', '__hash__', '__index__', '__init__', '__init_subclass__', '__int__', '__invert__', '__le__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'as_integer_ratio', 'bit_count', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']


### Exercício 3: Manipulação de variáveis de tipo string e explorando o uso de print

In [10]:
# Imprime os caracteres numéricos e seus códigos ASCII
for i in range(10):
    print(f"'{chr(48 + i)}' - {48 + i}")



'0' - 48
'1' - 49
'2' - 50
'3' - 51
'4' - 52
'5' - 53
'6' - 54
'7' - 55
'8' - 56
'9' - 57


In [11]:

# Imprime os caracteres numéricos e seus códigos ASCII usando ord()
for i in range(10):
    num_char = chr(ord('0') + i)  # Obtém o caractere numérico
    print(f"'{num_char}' - {ord(num_char)}")



'0' - 48
'1' - 49
'2' - 50
'3' - 51
'4' - 52
'5' - 53
'6' - 54
'7' - 55
'8' - 56
'9' - 57


> A função chr() converte um número inteiro em seu caractere correspondente na tabela ASCII, enquanto ord() faz o oposto, convertendo um caractere em seu valor numérico na tabela ASCII.

- Imprimindo o código numérico em octal e em hexadecimal.


In [12]:

# Imprime os caracteres numéricos e seus códigos ASCII usando ord()
print("Carac Deci Octa   Hex")
for i in range(10):
    num_char = chr(ord('0') + i)  # Obtém o caractere numérico
    print(f"'{num_char}' - {ord(num_char)} - {oct(ord(num_char))} - {hex(ord(num_char))}")



Carac Deci Octa   Hex
'0' - 48 - 0o60 - 0x30
'1' - 49 - 0o61 - 0x31
'2' - 50 - 0o62 - 0x32
'3' - 51 - 0o63 - 0x33
'4' - 52 - 0o64 - 0x34
'5' - 53 - 0o65 - 0x35
'6' - 54 - 0o66 - 0x36
'7' - 55 - 0o67 - 0x37
'8' - 56 - 0o70 - 0x38
'9' - 57 - 0o71 - 0x39


- Lendo um caractere qualquer e imprimindo no mesmo formato do código anterior


In [13]:
# Lê um caractere da entrada padrão
caractere = input("Digite um caractere: ")

# Obtém o código ASCII do caractere inserido
codigo_decimal = ord(caractere)
codigo_octal = oct(codigo_decimal)
codigo_hexadecimal = hex(codigo_decimal)
print("Carac Deci Octa   Hex")
# Imprime as representações do caractere
print(f"'{caractere}' - {ord(caractere)} - {oct(ord(caractere))} - {hex(ord(caractere))}")


Digite um caractere: d
Carac Deci Octa   Hex
'd' - 100 - 0o144 - 0x64


-  Caracteres especiais, ‘ç’ e ‘ã’


In [14]:
# Caracteres especiais 'ç' e 'ã'
caracteres_especiais = ['ç', 'ã']
print("Carac Deci Octa   Hex")
for caractere in caracteres_especiais:
    print(f"'{caractere}' - {ord(caractere)} - {oct(ord(caractere))} - {hex(ord(caractere))}")


Carac Deci Octa   Hex
'ç' - 231 - 0o347 - 0xe7
'ã' - 227 - 0o343 - 0xe3


### Exercício 4: Manipulação de variáveis de tipo string e explorando os métodos da classe.


In [15]:
# Declare uma variável com seu nome completo
nomeCompleto = "Beatriz Pereira Aragão"

In [16]:
# Separe o nome do sobrenome
nome, sobrenome = nomeCompleto.split(maxsplit=1)

> split(maxsplit=1) divide a string nomeCompleto em duas partes (logo após encontrar o primeiro espaço após a primeira string): a primeira palavra como o nome e o restante como o sobrenome.

In [17]:
# Verifique qual das variáveis antecede a outra na ordem alfabética
if nome < sobrenome:
    print(f"{nome} vem antes de {sobrenome} na ordem alfabética")
else:
    print(f"{sobrenome} vem antes de {nome} na ordem alfabética")




Beatriz vem antes de Pereira Aragão na ordem alfabética


In [18]:
# Verifique a quantidade de caracteres de cada variável
print(f"O nome tem {len(nome)} caracteres e o sobrenome tem {len(sobrenome)} caracteres")

O nome tem 7 caracteres e o sobrenome tem 14 caracteres


In [19]:
# Verifique se seu nome é um palíndromo
nome_minusculo = nome.lower()  # Convertendo para minúsculas para ignorar diferenças de caixa
if nome_minusculo == nome_minusculo[::-1]:
    print(f"{nome} é um palíndromo!")
else:
    print(f"{nome} não é um palíndromo.")

Beatriz não é um palíndromo.


> [:: -1] seleciona todos os elementos de uma sequência, começando pelo primeiro elemento e indo até o último, mas com um passo negativo (-1), o que efetivamente inverte a ordem dos elementos.

### Exercício 5: Manipulação de variáveis de ponto flutuante, explorando as características e os limites..


- Operadores aritméticos

In [20]:
# Valores para os cálculos
a = 12.0
b = 3.0

# Adição
soma = a + b
print(f"Soma de {a} e {b} é igual a {soma}")

# Subtração
subtracao = a - b
print(f"Subtração de {a} por {b} é igual a {subtracao}")

# Multiplicação
multiplicacao = a * b
print(f"Multiplicação de {a} por {b} é igual a {multiplicacao}")

# Divisão
divisao = a / b
print(f"Divisão de {a} por {b} é igual a {divisao}")

# Módulo
modulo = a % b
print(f"O resto da divisão de {a} por {b} é {modulo}")

# Exponenciação
exponenciacao = a ** b
print(f"{a} elevado a {b} é igual a {exponenciacao}")


Soma de 12.0 e 3.0 é igual a 15.0
Subtração de 12.0 por 3.0 é igual a 9.0
Multiplicação de 12.0 por 3.0 é igual a 36.0
Divisão de 12.0 por 3.0 é igual a 4.0
O resto da divisão de 12.0 por 3.0 é 0.0
12.0 elevado a 3.0 é igual a 1728.0


- Operadores aritméticos compostos

In [21]:
# Valores iniciais
a = 12.0
b = 3.0

# Adição com atribuição
x += y
print(f"Adição com atribuição: x += y resulta em x igual a {x}")

# Subtração com atribuição
x -= y
print(f"Subtração com atribuição: x -= y resulta em x igual a {x}")

# Multiplicação com atribuição
x *= y
print(f"Multiplicação com atribuição: x *= y resulta em x igual a {x}")

# Divisão com atribuição
x /= y
print(f"Divisão com atribuição: x /= y resulta em x igual a {x}")

# Módulo com atribuição
x %= y
print(f"Módulo com atribuição: x %= y resulta em x igual a {x}")

# Exponenciação com atribuição
x **= y
print(f"Exponenciação com atribuição: x **= y resulta em x igual a {x}")


Adição com atribuição: x += y resulta em x igual a 129
Subtração com atribuição: x -= y resulta em x igual a 65
Multiplicação com atribuição: x *= y resulta em x igual a 4160
Divisão com atribuição: x /= y resulta em x igual a 65.0
Módulo com atribuição: x %= y resulta em x igual a 1.0
Exponenciação com atribuição: x **= y resulta em x igual a 1.0


- A maior e a menor potência de 2 que pode ser representada com variáveis de ponto flutuante.


In [22]:
import sys
#maior potencia
maiorPotencia = 2 ** sys.float_info.max_exp
print(f"A maior potência de 2 representável é: {maiorPotencia}")

#menor
menorPotencia = 2 ** sys.float_info.min_exp
print(f"A menor potência de 2 representável é: {menorPotencia}")



A maior potência de 2 representável é: 179769313486231590772930519078902473361797697894230657273430081157732675805500963132708477322407536021120113879871393357658789768814416622492847430639474124377767893424865485276302219601246094119453082952085005768838150682342462881473913110540827237163350510684586298239947245938479716304835356329624224137216
A menor potência de 2 representável é: 4.450147717014403e-308


- As variveis são imutaveis

In [28]:
# Exemplo de obtenção do endereço de memória de uma variável
x = 42
y = x
print(f"Endereço de memória de 'x'={id(x)} e de y={id(y)}")
x +=1
print(f"Endereço de memória, após alteração, de 'x'={id(x)} e de y={id(y)}")


Endereço de memória de 'x'=140727022819400 e de y=140727022819400
Endereço de memória, após alteração, de 'x'=140727022819432 e de y=140727022819400


> Semelhante ao exemplo feito nos inteiros, mas agora fazendo o acompanhamento através do endereço que cada variavel esta apontando.

- Métodos estão disponíveis para as variáveis de ponto flutuante

In [24]:
print(dir(float))

['__abs__', '__add__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getformat__', '__getnewargs__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__int__', '__le__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__pos__', '__pow__', '__radd__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rmod__', '__rmul__', '__round__', '__rpow__', '__rsub__', '__rtruediv__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', 'as_integer_ratio', 'conjugate', 'fromhex', 'hex', 'imag', 'is_integer', 'real']


### Exercício 6: Manipulando listas.


> É possivel fatiar estruturas em python, chamamos de slices
 Seja L uma lista e usamos [p1:p2:p3], cada parametro fica responsavel por contrar uma parte dessa execução
 p1 = informa o inicio, 
 p2 = onde terminará e
 p3 = qual a quantidade de passos na leitura/separação
 
> Ao omitir um parametro, é usano os seu valores padrões:
p1 = primeiro elemento | p2 = ultimo elemento |p3 = passo de 1 em 1

- O que será impresso em cada caso: Sendo L = [1, 2, 3, 4, 5, 6, 7, 8, 9]


> print(L[::-1]) , a será exibido todos os elementos, mas de forma invertida 9,8,7,6,5,4,3,2,1;

>print(L[-1::]) , apenas o 9 será exibido pois o parametro de inicio está no -1 , ou seja o inicio é o ultimo elemento;

>print(L[:-1:]) , apenas o 9 não será exibido, pois o fim está para o penultimo elemento;

>print(L[::-2]) , os elementos de posição impares serão mostrados 1,3,5,7,9 na ordem inversa 9,7,3,5,1;

>print(L[-2::]) , exibe 9 e 8 , pois o inicio esta fixado no penultimo elemento;

>print(L[:-2:]) , não exibe 9 e 8 pois o fim esta fixado no antepenúltimo elemento;

###### Resposta

In [25]:
L = [1, 2, 3, 4, 5, 6, 7, 8, 9]
print(L[::-1])
print(L[-1::])
print(L[:-1:])
print(L[::-2])
print(L[-2::])
print(L[:-2:])

[9, 8, 7, 6, 5, 4, 3, 2, 1]
[9]
[1, 2, 3, 4, 5, 6, 7, 8]
[9, 7, 5, 3, 1]
[8, 9]
[1, 2, 3, 4, 5, 6, 7]


- Animal no zodiaco chines

In [26]:
print("Digite seu ano de nascimento: xxxx")
ano = int(input())
print(f"Seu signo chinês é: ", end="")

match (ano % 12):
    case 0:
        print("Macaco")
    case 1:
        print("Galo")
    case 2:
        print("Cão")
    case 3:
        print("Porco")
    case 4:
        print("Rato")
    case 5:
        print("Boi")
    case 6:
        print("Tigre")
    case 7:
        print("Coelho")
    case 8:
        print("Dragão")
    case 9:
        print("Serpente")
    case 10:
        print("Cavalo")
    case 11:
        print("Carneiro")
    case _:
        print("Valor inválido")

Digite seu ano de nascimento: xxxx
2001
Seu signo chinês é: Serpente
